packages/pds/src/api/com/atproto/account.ts

```diff
@@ -162,3 +145,37 @@ export default function (server: Server) {
     throw new InvalidRequestError('Not implemented')
   })
 }
+
+const ensureUsernameValid = (
+  username: string,
+  availableUserDomains: string[],
+): void => {
+  if (username.startsWith('did:')) {
+    throw new InvalidRequestError(
+      'Cannot register a username that starts with `did:`',
+      'InvalidUsername',
+    )
+  }
+  const supportedDomain = availableUserDomains.find((domain) =>
+    username.endsWith(domain),
+  )
+  if (!supportedDomain) {
+    throw new InvalidRequestError(
+      'Not a supported username domain',
+      'InvalidUsername',
+    )
+  }
+  const front = username.slice(0, username.length - supportedDomain.length)
+  if (front.length < 2) {
+    throw new InvalidRequestError('Username too short', 'InvalidUsername')
+  } else if (front.length > 20) {
+    throw new InvalidRequestError('Username too long', 'InvalidUsername')
+  }
+  const validChars = /^[a-zA-Z0-9-]*$/.test(front)
+  if (!validChars) {
+    throw new InvalidRequestError(
+      'Invalid characters in username',
+      'InvalidUsername',
+    )
+  }
```

Comment:
I wonder if it might be worth doing full domain validation here.  There are still plenty of ways to create invalid domains, including some esoteric ones such as placing a hyphen in the 3rd and 4th positions (conflicts with punycode) 🙃

Reply:
do you have a good canonical way of doing full domain validation? most of what im seeing online are esoteric regexes. which im not sure handle the esoteric cases

Reply:
Ah good question, I guess it is a little bit of a crapshoot.  Two that come to mind are these, though I see how we might not want to reach for them:
https://www.npmjs.com/package/is-valid-domain
https://github.com/hapijs/address/blob/d67f9ba02378155184b76ec34362a53410e92167/lib/domain.js

It's possible that this is more of a theoretical issue than a practical one at the moment.

Reply:
yea i came across that first one & it's deprecated (altho paul is a contributor on it 👀)

im fine with pullin in this hapi lib. seems simple/straightforward enough & handles the strange cases